File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] n1 = new int[nums.length];
        int[] n2 = new int[nums.length];

        n1[0] = 1;
        n2[nums.length - 1] = 1;

        for (int i = 1, j = nums.length - 2; i < nums.length; i++, j--) {
            n1[i] = nums[i - 1] * n1[i - 1];
            n2[j] = nums[j + 1] * n2[j + 1];
        }
        for (int i = 0; i < nums.length; i++) {
            n1[i] *= n2[i];
        }
        return n1;
    }
}

File: Data Structures & Algorithms/find-duplicate-integer/submission-5.kt
class Solution {
    fun findDuplicate(nums: IntArray): Int {
        var fast = 0
        var slow = 0
        do{
            fast = nums[nums[fast]]
            slow = nums[slow]
        }while(fast != slow)
        
        var begin = 0
        while(begin != slow){
            begin = nums[begin]
            slow = nums[slow]
        }
        return begin
    }
}
